### PR TITLE
ISPN-10429 Retry Hot Rod operations on socket timeout

### PR DESCRIPTION
**Pull request by tristantarrant**
https://issues.jboss.org/browse/ISPN-10429

### CODE REVIEW

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/HotRodOperation.java [hunk 1]
```diff
@@ -162,13 +161,22 @@ public boolean complete(T value) {
       return super.complete(value);
    }
 
-   public void scheduleTimeout(EventLoop eventLoop) {
+   public void scheduleTimeout(Channel channel) {
       assert timeoutFuture == null;
-      this.timeoutFuture = eventLoop.schedule(this, channelFactory.socketTimeout(), TimeUnit.MILLISECONDS);
+      this.timeoutFuture = channel.eventLoop().schedule(new TimeoutRunnable(channel), channelFactory.socketTimeout(), TimeUnit.MILLISECONDS);
```
Comment (rvansa) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/HotRodOperation.java]: Meh, that means a new allocation in each operation :-/ I wonder how much mess would be storing the channel directly in the `HotRodOperation`.

File: client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/EventLogListener.java [hunk 2]
```diff
@@ -71,21 +71,25 @@ public EventLogListener(RemoteCache<K, ?> remote) {
    @SuppressWarnings("unused")
    public void handleCreatedEvent(ClientCacheEntryCreatedEvent e) {
       createdEvents.add(e);
+      System.out.println("Created " + e);
```
Comment (rvansa) [inline client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/EventLogListener.java]: You'd probably rather use logging here?

Comment (tristantarrant) [reply client/hotrod-client/src/test/java/org/infinispan/client/hotrod/event/EventLogListener.java]: Sorry, leftover debugging

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/HotRodOperation.java (hunk 1 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/HotRodOperation.java]: Done

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java [hunk 3]
```diff
@@ -165,16 +164,16 @@ protected Throwable handleException(Throwable cause, ChannelHandlerContext ctx,
          if (address != null) {
             updateFailedServers(address);
          }
-         if (ctx != null) {
+         if (channel != null) {
             // We need to remove decoder even if we're about to close the channel
             // because otherwise we would be notified through channelInactive and we would retry (again).
-            if (ctx.pipeline().get(HeaderDecoder.NAME) != null) {
-               ctx.pipeline().remove(HeaderDecoder.NAME);
+            if (channel.pipeline().get(HeaderDecoder.NAME) != null) {
```
Comment (tristantarrant) [inline client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java]: @rvansa there is a problem here: event listener re-registrations are triggered by HeaderDecoder's channelInactive, but removing the handler from the pipeline, causes the subsequent channel.close() to skip that

Comment (rvansa) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java]: Not sure if I full understand, but you could invoke the same login as in channelInactive in `handlerRemoved`

**Comment by rvansa** [comment]
So, this was not covered by the testsuite in the past but I don't see any new test...

**Comment by tristantarrant** [comment]
The test which highlights this has just been merged in another pr: ResilienceIT will fail intermittently. But I will add a test here once I solve the listener

**Comment by tristantarrant** [comment]
This is now updated with a test.

File: client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java (hunk 3 above)
Comment (tristantarrant) [reply client/hotrod-client/src/main/java/org/infinispan/client/hotrod/impl/operations/RetryOnFailureOperation.java]: Done

**Comment by ryanemerson** [comment]
Thanks @tristantarrant

**Pull request merged**